File: 0009_is_palindrome/is_palindrome.cpp
#include <iostream>

using namespace std;

/**
 * Time: O(n)
 * Space: O(1)
 */
bool is_palindrome(int x)
{
    if (x < 0 || isFactorOfTen(x)) {
        return false;
    }
    int reverse = 0;
    while (x > reverse)
    {
        reverse = (reverse * 10) + x % 10;
        x /= 10;
    }
    return x == reverse || x == (reverse / 10);
}

bool isFactorOfTen(int x) {
    return x && x % 10 == 0;
}

File: 0009_is_palindrome/is_palindrome.js
/**
 * Time:    O(n)
 * Space:   O(1)
 * @param   {number}    x
 * @return  {boolean}
 */
const isPalindrome = (x) => {
    if (x < 0 || isFactorOfTen(x)) {
        return false
    }
    let reverse = 0
    while (x > reverse) {
        reverse = reverse * 10 + (x % 10)
        x = Math.floor(x / 10)
    }
    return x === reverse || x === Math.floor(reverse / 10)
}

const isFactorOfTen = (x) => x && x % 10 == 0
